sdk-cpp/src/internal/com/download_impl.cpp

```diff
@@ -186,109 +209,134 @@ class DOStatusCallback :
+    RETURN_RES_IF_FAILED(_codeInit);
+    return _spDownload->Start(nullptr);
 }
 
-void CDownloadImpl::Finalize()
+int32_t CDownloadImpl::Finalize() noexcept
 {
-    throw_if_fail(_spDownload->Finalize());
+    RETURN_RES_IF_FAILED(_codeInit);
+    return _spDownload->Finalize();
 }
 
-void CDownloadImpl::Abort()
+int32_t CDownloadImpl::Abort() noexcept
 {
-    throw_if_fail(_spDownload->Abort());
+    RETURN_RES_IF_FAILED(_codeInit);
+    return _spDownload->Abort();
 }
 
-void CDownloadImpl::SetCallback(const download_property_value::status_callback_t& callback, download& download)
+int32_t CDownloadImpl::GetStatus(msdo::download_status& status) noexcept
 {
-    Microsoft::WRL::ComPtr<DOStatusCallback> spCallback;
-    throw_if_fail(MakeAndInitialize<DOStatusCallback>(&spCallback, callback, download));
-
-    VARIANT vtCallback;
-    VariantInit(&vtCallback);
-    V_VT(&vtCallback) = VT_UNKNOWN;
-    V_UNKNOWN(&vtCallback) = spCallback.Get();
-    spCallback.Get()->AddRef();
-    const auto hr = _spDownload->SetProperty(ConvertToComProperty(msdo::download_property::callback_interface), &vtCallback);
-    VariantClear(&vtCallback);
-    throw_if_fail(hr);
+    RETURN_RES_IF_FAILED(_codeInit);
+    DO_DOWNLOAD_STATUS comStatus;
+    RETURN_RES_IF_FAILED(_spDownload->GetStatus(&comStatus));
+    status = ConvertFromComStatus(comStatus);
+    return S_OK;
 }
 
-msdo::download_status CDownloadImpl::GetStatus()
+int32_t CDownloadImpl::GetProperty(msdo::download_property key, msdo::download_property_value& value) noexcept
 {
-    DO_DOWNLOAD_STATUS status;
-    throw_if_fail(_spDownload->GetStatus(&status));
-    return ConvertFromComStatus(status);
+    return _GetPropertyHelper(key, value);
 }
 
-msdo::download_property_value CDownloadImpl::GetProperty(msdo::download_property key)
+int32_t CDownloadImpl::SetProperty(msdo::download_property key, const msdo::download_property_value& val) noexcept
 {
-    return _GetPropertyHelper(key);
+    RETURN_RES_IF_FAILED(_codeInit);
+
+    assert(key != msdo::download_property::callback_interface);
+    return _SetPropertyHelper(*_spDownload.Get(), key, val);
 }
 
-void CDownloadImpl::SetProperty(msdo::download_property key, const msdo::download_property_value& val)
+int32_t CDownloadImpl::SetCallback(const download_property_value::status_callback_t& callback, download& download) noexcept
 {
-    assert(key != msdo::download_property::callback_interface);
-    _SetPropertyHelper(*_spDownload.Get(), key, val);
+    RETURN_RES_IF_FAILED(_codeInit);
+
+    Microsoft::WRL::ComPtr<DOStatusCallback> spCallback;
+    RETURN_RES_IF_FAILED(MakeAndInitialize<DOStatusCallback>(&spCallback, callback, download));
+
+    VARIANT vtCallback;
+    VariantInit(&vtCallback);
+    V_VT(&vtCallback) = VT_UNKNOWN;
+    V_UNKNOWN(&vtCallback) = spCallback.Get();
+    spCallback.Get()->AddRef();
+    DODownloadProperty prop;
+    RETURN_RES_IF_FAILED(ConvertToComProperty(msdo::download_property::callback_interface, prop));
```

Comment:
>RETURN_RES_IF_FAILED [](http://example.com/codeflow?start=4&length=20)

This shouldn't fail, yes? Either convert to assert (with ifdef DEBUG, else) or just don't check the return value.

Reply:
Also remember that when we return fast like this we should perform cleanup like VariantClear before returning.